nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/AbstractCSVRecordReader.java

```diff
@@ -158,4 +180,46 @@ protected String trim(String value) {
     public RecordSchema getSchema() {
         return schema;
     }
+
+    /**
+     * This method searches using the specified Reader character-by-character until the
+     * record separator is found.
+     * @param reader the Reader providing the input
+     * @param recordSeparator the String specifying the end of a record in the input
+     * @throws IOException if an error occurs during reading, including not finding the record separator in the input
+     */
+    protected void readNextRecord(Reader reader, String recordSeparator) throws IOException {
+        int indexIntoSeparator = 0;
+        int recordSeparatorLength = recordSeparator.length();
+        int code = reader.read();
+        while (code != -1) {
+            char nextChar = (char)code;
+            if (recordSeparator.charAt(indexIntoSeparator) == nextChar) {
+                if (++indexIntoSeparator == recordSeparatorLength) {
+                    // We have matched the separator, return the string built so far
+                    return;
+                }
+            } else {
+                // The character didn't match the expected one in the record separator, reset the separator matcher
+                // and check if it is the first character of the separator.
+                indexIntoSeparator = 0;
+                if (recordSeparator.charAt(indexIntoSeparator) == nextChar) {
+                    // This character is the beginning of the record separator, keep it
+                    if (++indexIntoSeparator == recordSeparatorLength) {
+                        // We have matched the separator, return the string built so far
+                        return;
+                    }
+                }
+            }
```

Comment:
Instead of all the comments here, make a method which conveys the intention of what you are trying to accomplish.


```
char nextChar = (char)code;
if (recordSeparator.charAt(indexIntoSeparator) == nextChar) {
    ++indexIntoSeparator;
    if (hasReachedEndOfRecord(indexIntoSeparator, recordSeparatorLength) ) {
        return;
    }
} else {
    // The character didn't match the expected one in the record separator, reset the separator matcher
    // and check if it is the first character of the separator.
    indexIntoSeparator = 0;
    if (recordSeparator.charAt(indexIntoSeparator) == nextChar) {
        ++indexIntoSeparator;
        if (hasReachedEndOfRecord(indexIntoSeparator, recordSeparatorLength) ) {
         return;
        }
    }
}

private boolean hasReachedEndOfRecord(int indexIntoSeparator, int recordSeparatorLength) {
        return indexIntoSeparator == recordSeparatorLength;
}
```